apiserver/internal/handlers/objects/objects.go

```diff
@@ -8,391 +8,102 @@ import (
-	stateGetter              StateGetter
-	applicationServiceGetter ApplicationServiceGetter
-	makeCharmURL             CharmURLMakerFunc
-}
-
-// NewObjectsCharmHTTPHandler returns a new ObjectsCharmHTTPHandler.
-func NewObjectsCharmHTTPHandler(
+// NewObjectsHTTPHandler returns a new ObjectsHTTPHandler.
+func NewObjectsHTTPHandler(
 	stateGetter StateGetter,
-	applicationServiceGetter ApplicationServiceGetter,
-	charmURLMaker CharmURLMakerFunc,
-) http.Handler {
-	return &ObjectsCharmHTTPHandler{
-		stateGetter:              stateGetter,
-		applicationServiceGetter: applicationServiceGetter,
-		makeCharmURL:             charmURLMaker,
+	objectStoreGetter ObjectStoreServiceGetter,
+) *ObjectsHTTPHandler {
+	return &ObjectsHTTPHandler{
+		stateGetter:       stateGetter,
+		objectStoreGetter: objectStoreGetter,
 	}
 }
 
 // ServeHTTP implements the http.Handler interface.
-func (h *ObjectsCharmHTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+func (h *ObjectsHTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	var err error
 	switch r.Method {
 	case "GET":
 		err = h.ServeGet(w, r)
 		if err != nil {
-			err = errors.Errorf("cannot retrieve charm: %w", err)
-		}
-	case "PUT":
-		err = h.ServePut(w, r)
-		if err != nil {
-			err = errors.Errorf("cannot upload charm: %w", err)
+			err = errors.Errorf("cannot retrieve object: %w", err)
 		}
 	default:
 		http.Error(w, fmt.Sprintf("http method %s not implemented", r.Method), http.StatusNotImplemented)
 		return
 	}
 
-	if err != nil {
-		if err := sendJSONError(w, errors.Capture(err)); err != nil {
-			logger.Errorf(r.Context(), "%v", errors.Errorf("cannot return error to user: %w", err))
-		}
+	if err == nil {
+		return
+	}
+
+	if err := sendJSONError(w, errors.Capture(err)); err != nil {
+		logger.Errorf(r.Context(), "%v", errors.Errorf("cannot return error to user: %w", err))
 	}
 }
 
 // ServeGet serves the GET method for the S3 API. This is the equivalent of the
 // `GetObject` method in the AWS S3 API.
-func (h *ObjectsCharmHTTPHandler) ServeGet(w http.ResponseWriter, r *http.Request) error {
-	applicationService, err := h.applicationServiceGetter.Application(r)
+func (h *ObjectsHTTPHandler) ServeGet(w http.ResponseWriter, r *http.Request) error {
+	service, err := h.objectStoreGetter.ObjectStore(r)
 	if err != nil {
 		return errors.Capture(err)
 	}
 
 	query := r.URL.Query()
-	_, charmSha256Prefix, err := splitNameAndSHAFromQuery(query)
-	if err != nil {
-		return err
+	sha256 := query.Get(":object")
+	if sha256 == "" {
+		return jujuerrors.BadRequestf("missing object sha256")
 	}
 
-	reader, err := applicationService.GetCharmArchiveBySHA256Prefix(r.Context(), charmSha256Prefix)
-	if errors.Is(err, applicationerrors.CharmNotFound) {
-		return jujuerrors.NotFoundf("charm")
+	reader, readerSize, err := service.GetBySHA256(r.Context(), sha256)
+	if errors.Is(err, objectstoreerrors.ErrNotFound) {
+		return jujuerrors.NotFoundf("object: %s", sha256)
```

Comment:
`sha256` has to be a full SHA here right? Do you think it's worth asserting the length is correct?

I know the metadata service performs that check itself. Might be worth doing here as well though? What d oyou think?

Reply:
That's done in the service. It's done early on in the service, so it's a bit overkill.